sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonStreamingSourceRunner.scala

```diff
@@ -199,4 +223,30 @@ class PythonStreamingSourceRunner(
         logError("Exception when trying to kill worker", e)
     }
   }
+
+  private val allocator = ArrowUtils.rootAllocator.newChildAllocator(
+    s"stream reader for $pythonExec", 0, Long.MaxValue)
+
+  def readArrowRecordBatches(): Iterator[InternalRow] = {
+    assert(dataIn.readInt() == SpecialLengths.START_ARROW_STREAM)
+    val reader = new ArrowStreamReader(dataIn, allocator)
+    val root = reader.getVectorSchemaRoot()
+    // When input is empty schema can't be read.
+    val schema = ArrowUtils.fromArrowSchema(root.getSchema())
+    assert(schema == outputSchema)
+
+    val vectors = root.getFieldVectors().asScala.map { vector =>
+      new ArrowColumnVector(vector)
+    }.toArray[ColumnVector]
+    val rows = ArrayBuffer[InternalRow]()
```

Comment:
We are going to buffer all the rows in memory here? Can we create this iterator lazily to avoid buffering all data from Python source?

Reply:
We can't do lazy initialization here because we need to send the data from python process to JVM, the communication is synchronous.

Reply:
I imagine there may be still a way to avoid materializing all rows at once (e.g. per arrow batch) but I don't concern too much about it as we know simple data source isn't intended to handle a huge amount of data.

Reply:
If we call putIterator here we should be able to avoid materializing all rows at once in scala side, but it doesn't matter that much as we already materialize all rows in python side.